subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -112,23 +112,26 @@ private void updateWatchRegistry(Consumer<FileWatcherRegistry> updateFunction, R
 
     @Override
     public void buildRootDirectoryAdded(File buildRootDirectory) {
-        synchronized (buildRootDirectoriesForWatching) {
-            buildRootDirectoriesForWatching.add(buildRootDirectory);
-            updateWatchRegistry(watchRegistry -> watchRegistry.getFileWatcherUpdater().updateProjectRootDirectories(buildRootDirectoriesForWatching));
+        synchronized (rootProjectDirectoriesForWatching) {
+            rootProjectDirectoriesForWatching.add(buildRootDirectory);
+            updateWatchRegistry(watchRegistry -> watchRegistry.getFileWatcherUpdater().updateRootProjectDirectories(rootProjectDirectoriesForWatching));
         }
     }
 
     @Override
     public void beforeBuildFinished(boolean watchingEnabled) {
-        synchronized (buildRootDirectoriesForWatching) {
-            buildRootDirectoriesForWatching.clear();
+        synchronized (rootProjectDirectoriesForWatching) {
+            rootProjectDirectoriesForWatching.clear();
         }
         if (watchingEnabled) {
             getRoot().update(currentRoot -> {
                 buildRunning = false;
                 producedByCurrentBuild.set(DefaultFileHierarchySet.of());
                 SnapshotHierarchy newRoot = removeSymbolicLinks(currentRoot);
                 newRoot = handleWatcherRegistryEvents(newRoot, "for current build");
+                if (watchRegistry != null) {
+                    newRoot = handleWatcherChangeErrors(newRoot, () -> watchRegistry.getFileWatcherUpdater().buildFinished());
```

Comment:
IIUC, the purpose of `handleWatcherChangeErrors()` is to run the given action, and handle any watcher errors arising from it. But the name suggests to me that the purpose of this method is to handle errors: there is no conditionality captured. I think something like `withWatcherChangeErrorHandling()` would capture the purpose better. Not feeling very strongly about it, but...